File: frontend/de/unika/ipd/grgen/ast/decl/DeclNode.java
/*
 * GrGen: graph rewrite generator tool -- release GrGen.NET 7.0
 * Copyright (C) 2003-2024 Universitaet Karlsruhe, Institut fuer Programmstrukturen und Datenorganisation, LS Goos; and free programmers
 * licensed under LGPL v3 (see LICENSE.txt included in the packaging of this file)
 * www.grgen.net
 */

/**
 * @author Sebastian Hack
 */

package de.unika.ipd.grgen.ast.decl;

import java.awt.Color;

import de.unika.ipd.grgen.ast.BaseNode;
import de.unika.ipd.grgen.ast.DeclaredCharacter;
import de.unika.ipd.grgen.ast.IdentNode;
import de.unika.ipd.grgen.ast.decl.pattern.EdgeDeclNode;
import de.unika.ipd.grgen.ast.model.type.ArbitraryEdgeTypeNode;
import de.unika.ipd.grgen.ast.model.type.DirectedEdgeTypeNode;
import de.unika.ipd.grgen.ast.model.type.UndirectedEdgeTypeNode;
import de.unika.ipd.grgen.ast.type.TypeNode;
import de.unika.ipd.grgen.ir.Entity;

/**
 * Base class for all AST nodes representing declarations.
 * children: IDENT:IdentNode TYPE:
 */
public abstract class DeclNode extends BaseNode implements DeclaredCharacter
{
	static {
		setName(DeclNode.class, "declaration");
	}

	public IdentNode ident;

	public BaseNode typeUnresolved;

	/** An invalid declaration. */
	private static final DeclNode invalidDecl = new InvalidDeclNode(IdentNode.getInvalid());

	/** Get an invalid declaration. */
	public static final DeclNode getInvalid()
	{
		return invalidDecl;
	}

	/** Get an invalid declaration for an IdentNode. */
	public static final DeclNode getInvalid(IdentNode id)
	{
		return new InvalidDeclNode(id);
	}

	/**
	 * Create a new declaration node
	 * @param n The identifier that is declared
	 * @param t The type with which it is declared
	 */
	protected DeclNode(IdentNode n, BaseNode t)
	{
		super(n.getCoords());
		n.setDecl(this);
		this.ident = n;
		becomeParent(this.ident);
		this.typeUnresolved = t;
		becomeParent(this.typeUnresolved);
	}

	/** @return The ident node of the declaration */
	public IdentNode getIdentNode()
	{
		return ident;
	}

	/** @return The type node of the declaration */
	public abstract TypeNode getDeclType();

	/** @see de.unika.ipd.grgen.ast.DeclaredCharacter#getDecl() */
	@Override
	public DeclNode getDecl()
	{
		return this;
	}

	/** @see de.unika.ipd.grgen.util.GraphDumpableNode#getNodeColor() */
	@Override
	public Color getNodeColor()
	{
		return Color.BLUE;
	}

	public final String emptyWhenAnonymous(String str)
	{
		return getIdentNode().getCurrOcc().isAnonymous() ? "" : str;
	}

	public final String emptyWhenAnonymousPostfix(String prefix)
	{
		return getIdentNode().getCurrOcc().isAnonymous() ? "" : prefix + getIdentNode();
	}

	public final String emptyWhenAnonymousInParenthesis(String prefix)
	{
		return getIdentNode().getCurrOcc().isAnonymous() ? "" : prefix + "(" + getIdentNode() + ")";
	}

	public final String dotOrArrowWhenAnonymous()
	{
		if(getIdentNode().getCurrOcc().isAnonymous() && this instanceof EdgeDeclNode)
		{
			EdgeDeclNode edge = (EdgeDeclNode)this;
			if(edge.getDeclType() instanceof ArbitraryEdgeTypeNode)
				return "?--?";
			else if(edge.getDeclType() instanceof DirectedEdgeTypeNode)
				return "-->";
			else if(edge.getDeclType() instanceof UndirectedEdgeTypeNode)
				return "--";
		}
		return getIdentNode().getCurrOcc().isAnonymous() ? "." : getIdentNode().toString();
	}

	public Entity getEntity()
	{
		return checkIR(Entity.class);
	}

	//@Override
	//protected IR constructIR() {
	//	Type type = getDeclType().checkIR(Type.class);
	//	return new Entity("entity", getIdentNode().getIdent(), type, false);
	//}

	public static String getKindStr()
	{
		return "declaration";
	}

	@Override
	public String toString()
	{
		return ident.toString();
	}
}

File: frontend/de/unika/ipd/grgen/ast/decl/DeclNode.cs
/*
 * GrGen: graph rewrite generator tool -- release GrGen.NET 8.1
 * Copyright (C) 2003-2026 Universitaet Karlsruhe, Institut fuer Programmstrukturen und Datenorganisation, LS Goos; and free programmers
 * licensed under LGPL v3, some components/parts use different licenses (see LICENSE.txt included in the packaging of this file)
 * www.grgen.de / www.grgen.net
 */

/// <summary>
/// @author Sebastian Hack
/// </summary>

namespace de.unika.ipd.grgen.ast.decl
{

	using BaseNode = de.unika.ipd.grgen.ast.BaseNode;
	using DeclaredCharacter = de.unika.ipd.grgen.ast.DeclaredCharacter;
	using IdentNode = de.unika.ipd.grgen.ast.IdentNode;
	using EdgeDeclNode = de.unika.ipd.grgen.ast.decl.pattern.EdgeDeclNode;
	using ArbitraryEdgeTypeNode = de.unika.ipd.grgen.ast.model.type.ArbitraryEdgeTypeNode;
	using DirectedEdgeTypeNode = de.unika.ipd.grgen.ast.model.type.DirectedEdgeTypeNode;
	using UndirectedEdgeTypeNode = de.unika.ipd.grgen.ast.model.type.UndirectedEdgeTypeNode;
	using TypeNode = de.unika.ipd.grgen.ast.type.TypeNode;
	using Entity = de.unika.ipd.grgen.ir.Entity;
	using Color = de.unika.ipd.grgen.util.Color;

	/// <summary>
	/// Base class for all AST nodes representing declarations.
	/// children: IDENT:IdentNode TYPE:
	/// </summary>
	public abstract class DeclNode : BaseNode, DeclaredCharacter
	{
		static DeclNode()
		{
			SetClassName(typeof(DeclNode), "declaration");
			invalidDecl = new InvalidDeclNode(IdentNode.Invalid);
		}

		public IdentNode ident;

		public BaseNode typeUnresolved;

		/// <summary>
		/// An invalid declaration. </summary>
		private static readonly DeclNode invalidDecl;

		/// <summary>
		/// Get an invalid declaration. </summary>
		public static DeclNode Invalid
		{
			get
			{
				return invalidDecl;
			}
		}

		/// <summary>
		/// Get an invalid declaration for an IdentNode. </summary>
		public static DeclNode GetInvalid(IdentNode id)
		{
			return new InvalidDeclNode(id);
		}

		/// <summary>
		/// Create a new declaration node </summary>
		/// <param name="n"> The identifier that is declared </param>
		/// <param name="t"> The type with which it is declared </param>
		protected internal DeclNode(IdentNode n, BaseNode t)
			: base(n.Coords)
		{
			n.Decl = this;
			this.ident = n;
			BecomeParent(this.ident);
			this.typeUnresolved = t;
			BecomeParent(this.typeUnresolved);
		}

		/// <returns> The ident node of the declaration </returns>
		public virtual IdentNode Ident
		{
			get
			{
				return ident;
			}
		}

		/// <returns> The type node of the declaration </returns>
		public abstract TypeNode DeclType {get;}

		/// <seealso cref="de.unika.ipd.grgen.ast.DeclaredCharacter.getDecl() "/>
		public virtual DeclNode Decl
		{
			get
			{
				return this;
			}
		}

		/// <seealso cref="de.unika.ipd.grgen.util.GraphDumpableNode.getNodeColor() "/>
		public override Color NodeColor
		{
			get
			{
				return Color.BLUE;
			}
		}

		public string EmptyWhenAnonymous(string str)
		{
			return Ident.CurrOcc.IsAnonymous() ? "" : str;
		}

		public string EmptyWhenAnonymousPostfix(string prefix)
		{
			return Ident.CurrOcc.IsAnonymous() ? "" : prefix + Ident;
		}

		public string EmptyWhenAnonymousInParenthesis(string prefix)
		{
			return Ident.CurrOcc.IsAnonymous() ? "" : prefix + "(" + Ident + ")";
		}

		public string DotOrArrowWhenAnonymous()
		{
			if(Ident.CurrOcc.IsAnonymous() && this is EdgeDeclNode)
			{
				EdgeDeclNode edge = (EdgeDeclNode)this;
				if(edge.DeclType is ArbitraryEdgeTypeNode)
					return "?--?";
				else if(edge.DeclType is DirectedEdgeTypeNode)
					return "-->";
				else if(edge.DeclType is UndirectedEdgeTypeNode)
					return "--";
			}
			return Ident.CurrOcc.IsAnonymous() ? "." : Ident.ToString();
		}

		public virtual Entity IREntity
		{
			get
			{
				return CheckIR<Entity>(typeof(Entity));
			}
		}

		public static string KindStr
		{
			get
			{
				return "declaration";
			}
		}

		public override string ToString()
		{
			return ident.ToString();
		}
	}

}
